com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensor.cs

```diff
@@ -279,27 +290,58 @@ public RayPerceptionOutput RayPerceptionOutput
             get { return m_RayPerceptionOutput; }
         }
 
-        void SetNumObservations(int numObservations)
+        /// <summary>
+        /// The observation size per ray.
+        /// Override this method for custom observations.
+        /// </summary>
+        public virtual int GetObservationSizePerRay()
+        {
+            return RayPerceptionInput.DetectableTags.Count + 2;
+        }
+
+        /// <summary>
+        /// The number of rays in the sensor.
+        /// Override this method for custom observations.
+        /// </summary>
+        public int GetNumberOfRays()
+        {
+            return RayPerceptionInput.Angles.Count;
+        }
+
+        void SetNumObservations(int observationsSize, int numRays)
         {
-            m_ObservationSpec = ObservationSpec.Vector(numObservations);
-            m_Observations = new float[numObservations];
+            m_ObservationSpec = ObservationSpec.Vector(observationsSize * numRays);
+            m_Observations = new float[observationsSize * numRays];
+            m_SingleRayObservations = new float[observationsSize];
         }
 
         internal void SetRayPerceptionInput(RayPerceptionInput rayInput)
         {
             // Note that change the number of rays or tags doesn't directly call this,
             // but changing them and then changing another field will.
-            if (m_RayPerceptionInput.OutputSize() != rayInput.OutputSize())
+            var oldObservationSize = GetObservationSizePerRay();
+            m_RayPerceptionInput = rayInput;
+            if (GetObservationSizePerRay() != oldObservationSize)
```

Comment:
This change misses the case where the observation size per ray stays the same, but the number of rays changes.

Reply:
fixed